extensions/storages/source/deployment.ts

```diff
@@ -16,50 +19,53 @@ export function deployment (instances: Instance[], annotation: Annotation): Depe
 }
 
 function validate (instances: Instance[], annotation: Annotation): void {
-  if (annotation === undefined)
-    throw new Error('Storages annotation is required by: ' +
-      `'${instances.map((i) => i.component.locator.id).join("', '")}'`)
-
-  for (const instance of instances)
-    contains(instance, annotation)
+  assert.ok(annotation !== undefined,
+    `Storages annotation is required by: '${instances
+      .map((i) => i.component.locator.id)
+      .join("', '")}'`)
 
-  for (const ref of Object.values(annotation)) {
-    const url = new URL(ref)
+  for (const instance of instances) contains(instance, annotation)
 
-    if (!(url.protocol in providers))
-      throw new Error(`Unknown storage provider '${url.protocol}'`)
-  }
+  for (const { provider } of Object.values(annotation))
+    validateProviderId(provider)
 }
 
 function contains (instance: Instance, annotation: Annotation): void {
   for (const name of instance.manifest)
-    if (!(name in annotation))
-      throw new Error(`Missing '${name}' storage annotation ` +
+    assert.ok(name in annotation,
+      `Missing '${name}' storage annotation ` +
         `declared in '${instance.component.locator.id}'`)
 }
 
-function getSecrets (annotation: Annotation): Variable[] {
+export function validateProviderId (id: string | undefined): asserts id is keyof typeof providers {
+  assert.ok(typeof id === 'string' && id in providers, `Unknown storage provider '${id}'`)
+}
+
+function getSecrets (annotation: Readonly<Annotation>): Variable[] {
   const secrets: Variable[] = []
 
-  for (const [storage, ref] of Object.entries(annotation)) {
-    const url = new URL(ref)
-    const Provider = providers[url.protocol]
+  for (const [storage, props] of Object.entries(annotation)) {
+    const providerId = props.provider
+
+    validateProviderId(providerId)
```

Comment:
This validation has already been done at this point.

Reply:
It's a Typescript typeguard in this context.

Reply:
Ok, I will not argue here.

But this kind of practice will not be acceptable on hot code paths.

We cannot run actual code in production that only exists to satisfy meta-code requirements on developers machines.

Reply:
Fixed in https://github.com/toa-io/toa/pull/493/commits/6cc7026fb95dd765addabc2d3d30712d00455fe8